Cached Chunker panics on nil IPLD link

Observed on head of https://github.com/filecoin-project/boost/tree/nonsense/piece-store

Cc @dirkmc @nonsense

```
panic: interface conversion: datamodel.Link is nil, not cidlink.Link [recovered]
        panic: interface conversion: datamodel.Link is nil, not cidlink.Link [recovered]
        panic: interface conversion: datamodel.Link is nil, not cidlink.Link

goroutine 347 [running]:
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.func1()
        /root/code/pkg/mod/go.opentelemetry.io/otel/sdk@v1.9.0/trace/span.go:363 +0x2a
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc0153aca80, {0x0, 0x0, 0xf6a9ed70fb50a084?})
        /root/code/pkg/mod/go.opentelemetry.io/otel/sdk@v1.9.0/trace/span.go:402 +0x8ee
panic({0x3a42dc0, 0xc01df7fbc0})
        /usr/local/go/src/runtime/panic.go:884 +0x212
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.func1()
        /root/code/pkg/mod/go.opentelemetry.io/otel/sdk@v1.9.0/trace/span.go:363 +0x2a
go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc0153acd80, {0x0, 0x0, 0x0?})
        /root/code/pkg/mod/go.opentelemetry.io/otel/sdk@v1.9.0/trace/span.go:402 +0x8ee
panic({0x3a42dc0, 0xc01df7fbc0})
        /usr/local/go/src/runtime/panic.go:884 +0x212
github.com/filecoin-project/index-provider/engine/chunker.dsKey({0x0?, 0x0?})
        /root/code/pkg/mod/github.com/filecoin-project/index-provider@v0.8.1/engine/chunker/cached_chunker.go:219 +0x77
github.com/filecoin-project/index-provider/engine/chunker.(*CachedEntriesChunker).dsRootPrefixedKey(0xc013b5f200?, {0x0?, 0x0?})
        /root/code/pkg/mod/github.com/filecoin-project/index-provider@v0.8.1/engine/chunker/cached_chunker.go:506 +0x2d
github.com/filecoin-project/index-provider/engine/chunker.(*CachedEntriesChunker).Chunk(0xc013b5f200, {0x4675e00, 0xc013981180}, {0x465d440, 0xc01dff9a60})
        /root/code/pkg/mod/github.com/filecoin-project/index-provider@v0.8.1/engine/chunker/cached_chunker.go:277 +0xbe6
github.com/filecoin-project/index-provider/engine.(*Engine).mkLinkSystem.func1({{0x4675e00, 0xc013981180}, {{0x0, 0x0, 0x0}}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}}, ...)
        /root/code/pkg/mod/github.com/filecoin-project/index-provider@v0.8.1/engine/linksystem.go:106 +0xab4
github.com/ipfs/go-graphsync/responsemanager/queryexecutor.(*QueryExecutor).loadBlock(_, {_, _}, {0x0, {0x4675e00, 0xc022f48040}, {0x4685fc0, 0xc010171800}, {{{0xc025c683f0, 0x15}}, ...}, ...}, ...)
        /root/code/pkg/mod/github.com/ipfs/go-graphsync@v0.13.1/responsemanager/queryexecutor/queryexecutor.go:222 +0x1e6
github.com/ipfs/go-graphsync/responsemanager/queryexecutor.(*QueryExecutor).runTraversal(0xa5cc82?, {0x4675ea8, 0xc023adc630}, {0xc022f348a0, 0x26}, {0x0, {0x4675e00, 0xc022f48040}, {0x4685fc0, 0xc010171800}, ...})
        /root/code/pkg/mod/github.com/ipfs/go-graphsync@v0.13.1/responsemanager/queryexecutor/queryexecutor.go:203 +0x45c
github.com/ipfs/go-graphsync/responsemanager/queryexecutor.(*QueryExecutor).executeQuery(0xc0026816c0?, {0x4675ea8?, 0xc023adc630?}, {0xc022f348a0?, 0xc011f83a50?}, {0x0, {0x4675e00, 0xc022f48040}, {0x4685fc0, 0xc010171800}, ...})
        /root/code/pkg/mod/github.com/ipfs/go-graphsync@v0.13.1/responsemanager/queryexecutor/queryexecutor.go:121 +0x7e
github.com/ipfs/go-graphsync/responsemanager/queryexecutor.(*QueryExecutor).ExecuteTask(0xc011f05900, {0xc011f83f24?, 0x3579334d694d7877?}, {0xc022f348a0, 0x26}, 0x7d22677277366a01?)
        /root/code/pkg/mod/github.com/ipfs/go-graphsync@v0.13.1/responsemanager/queryexecutor/queryexecutor.go:105 +0x586
github.com/ipfs/go-graphsync/taskqueue.(*WorkerTaskQueue).worker(0xc011f3ae60, {0x4660b20, 0xc011f05900})
        /root/code/pkg/mod/github.com/ipfs/go-graphsync@v0.13.1/taskqueue/taskqueue.go:139 +0x3b5
created by github.com/ipfs/go-graphsync/taskqueue.(*WorkerTaskQueue).Startup
        /root/code/pkg/mod/github.com/ipfs/go-graphsync@v0.13.1/taskqueue/taskqueue.go:97 +0x3a
```

--- a/engine/chunker/cached_chunker.go
+++ b/engine/chunker/cached_chunker.go
@@ -248,7 +248,7 @@ func (ls *CachedEntriesChunker) Chunk(ctx context.Context, mhi provider.Multihas
 	// Intercept the links that are being stored.
 	// It is safe to swap the StorageWriteOpener, because:
 	//  - Chunk is the only place we expect to write to the linksystem, and
-	//  - calls to Chunk are syncronized using a lock.
+	//  - calls to Chunk are synchronized using a lock.
 	// It is also an efficient way to collecting all the links without having to traverse the dag
 	// from the root link, or make the EntriesChunker interface more complex.
 	ls.lsys.StorageWriteOpener = func(ctx linking.LinkContext) (io.Writer, linking.BlockWriteCommitter, error) {
@@ -267,6 +267,9 @@ func (ls *CachedEntriesChunker) Chunk(ctx context.Context, mhi provider.Multihas
 	root, err := ls.chunker.Chunk(ctx, mhi)
 	if err != nil {
 		return nil, err
+	} else if root == nil {
+		log.Debugw("multihash iterator returned no elements")
+		return nil, nil
 	}
 
 	// Store internal mappings for caching purposes.

--- a/engine/chunker/chunker.go
+++ b/engine/chunker/chunker.go
@@ -12,5 +12,6 @@ import (
 type EntriesChunker interface {
 	// Chunk chunks multihashes supplied by a given provider.MultihashIterator into a chain of
 	// schema.EntryChunk and returns the link of the chain root.
+	// If the given iterator has no elements, this function returns a nil link with no error.
 	Chunk(context.Context, provider.MultihashIterator) (ipld.Link, error)
 }

--- a/engine/chunker/hamt_chunker.go
+++ b/engine/chunker/hamt_chunker.go
@@ -75,14 +75,20 @@ func (h *HamtChunker) Chunk(ctx context.Context, iterator provider.MultihashIter
 	if err != nil {
 		return nil, err
 	}
+	var count int
 	for {
 		mh, err := iterator.Next()
 		if err == io.EOF {
+			// Iterator had no elements
+			if count == 0 {
+				return nil, nil
+			}
 			break
 		}
 		if err != nil {
 			return nil, err
 		}
+		count++
 		if err := ma.AssembleKey().AssignBytes(mh); err != nil {
 			return nil, err
 		}
@@ -90,6 +96,7 @@ func (h *HamtChunker) Chunk(ctx context.Context, iterator provider.MultihashIter
 			return nil, err
 		}
 	}
+	log.Debugw("finished iterating over multihash lister", "mhCount", count)
 	if err := ma.Finish(); err != nil {
 		return nil, err
 	}

--- a/engine/linksystem.go
+++ b/engine/linksystem.go
@@ -15,7 +15,12 @@ import (
 	"github.com/libp2p/go-libp2p/core/peer"
 )
 
-var errNoEntries = errors.New("no entries; see schema.NoEntries")
+var (
+	errNoEntries = errors.New("no entries; see schema.NoEntries")
+
+	// ErrEntriesLinkMismatch signals that the link generated from chunking the mulithashes returned by provider.MultihashLister does not match the previously generated link. This error is most likely caused by the lister returning inconsistent multihashes for the same key.
+	ErrEntriesLinkMismatch = errors.New("regenerated link from multihash lister did not match the original link; multihashes returned by the lister for the same key are not consistent")
+)
 
 // Creates the main engine linksystem.
 func (e *Engine) mkLinkSystem() ipld.LinkSystem {
@@ -109,10 +114,14 @@ func (e *Engine) mkLinkSystem() ipld.LinkSystem {
 			// Store the linked list entries in cache as we generate them.  We
 			// use the cache linksystem that stores entries in an in-memory
 			// datastore.
-			_, err = e.entriesChunker.Chunk(ctx, mhIter)
+			regeneratedLink, err := e.entriesChunker.Chunk(ctx, mhIter)
 			if err != nil {
 				log.Errorf("Error generating linked list from multihash lister: %s", err)
 				return nil, err
+			}
+			if regeneratedLink == nil || !c.Equals(regeneratedLink.(cidlink.Link).Cid) {
+				log.Errorw("Regeneration of entries link from multihash iterator did not match the original link. Check that multihash iterator consistently returns the same entries for the same key.", "want", lnk, "got", regeneratedLink)
+				return nil, ErrEntriesLinkMismatch
 			}
 		} else {
 			log.Debugw("Found cache entry for CID", "cid", c)

